File: leetcode/Java/0074-m-search-a-2d-matrix.java
/* 
 * Binary search
 */
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix == null || matrix.length == 0) {
            return false;
        }

        if (matrix[0] == null || matrix[0].length == 0) {
            return false;
        }

        int n = matrix.length;
        int m = matrix[0].length;
        int start = 0, end = n * m - 1;

        while (start + 1 < end) {
            int mid = start + (end - start) / 2;
            int row = mid / m;
            int col = mid % m;
            if (matrix[row][col] < target) {
                start = mid;
            } else {
                end = mid;
            }
        }

        if (matrix[start / m][start % m] == target) {
            return true;
        }

        if (matrix[end / m][end % m] == target) {
            return true;
        }

        return false;
    }
}

File: leetcode/C++/0074-m-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(const vector<vector<int>>& matrix, int target) {
        if (matrix.empty()) return false;
        const size_t m = matrix.size();
        const size_t n = matrix.front().size();

        int first = 0;
        int last = m * n;

        while (first < last) {
            int mid = first + (last - first) / 2;
            int value = matrix[mid / n][mid % n];

            if (value == target)
                return true;
            else if (value < target)
                first = mid + 1;
            else
                last = mid;
        }
        return false;
    }
};


class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if (matrix.empty()) return false;
        int m = matrix.size();
        int n = matrix[0].size();

        int i = m - 1, j = 0;
        while (i > -1 && j < n) {
            if (target == matrix[i][j])
                return true;
            else if (target < matrix[i][j])
                i--;
            else
                j++;
        }
        return false;
    }
};